### ISSUE TITLE
Ask for help about eval own trained model.

### ISSUE THREAD

**Issue by HelloWorld416** [autonomousvision/transfuser#110]
Why do I use the dataset （210G） you provided to train the model by myself, during test phase on the Carla , the ego car can't go straight on the road but just make a turn and run out of the road? In the training phase, we try to train 'only_wp'=1 and set --setting with both 'all' and '02_05_withheld'. I use one GPU 3090 to train the model and set epoch for 41, batch_size=32, other settings are default. I am very confused with this phenomenon and appreciate if you could provide some idea for me.

**Comment by Kait0** [comment]
Hard to tell what is going wrong without any additional information.
When you run the evaluation with the provided pre-trained weights is the model behaving normally then?
When you run your trained model, are the predicted waypoints also going off the road (or is there some issue with the controller)?
Did you make any modifications to the code?
What is the full training configuration that you used?
How long did the training take?
Does this also happen when you train with the default configuration?

**Comment by HelloWorld416** [comment]
1.When I use the provided pre-trained weights, the model behaves normally.
2.We print the ‘pred_wp’ from submission_agent.py, and the results are like:
[[-0.6865, -0.1091], [-0.0335, -0.2020], [0.6574, -0.2611], [1.3772, -0.3137]];
[[-0.6966, -0.1141], [-0.0542, -0.1943], [0.6270, -0.2508], [1.3374, -0.3006]]; 
[[-0.7083, -0.1091], [-0.0770, -0.1863], [0.5937, -0.2406], [1.2937, -0.2887]].
.........
But the car is running on straight road. The controller is you provide and I do not change.
3.The code is all you provided, I just delete some setting about multi GPUs for training.
4.The full training configuration is following:
{
  "id": "transfuser",
  "epochs": 41,
  "lr": 0.0001,
  "batch_size": 32,
  "start_epoch": 0,
  "setting": "all",
  "logdir": ".........",
  "load_file": null,
  "dataset_dir": "...........",
  "schedule": 1,
  "schedule_reduce_epoch_01": 10,
  "schedule_reduce_epoch_02": 20,
  "schedule_reduce_epoch_03": 30,
  "backbone": "transFuser",
  "image_architecture": "regnety_032",
  "lidar_architecture": "regnety_032",
  "use_velocity": 0,
  "n_layer": 4,
  "wp_only": 0,
  "use_target_point_image": 1,
  "use_point_pillars": 0,
  "val_every": 5,
  "no_bev_loss": 0,
  "sync_batch_norm": 0,
  "zero_redundancy_optimizer": 0
}
5.It takes an average of four and a half hours to train for one epoch, and about 7.7 days after 41 epochs.
6.I git clone this repo and download the dataset (210G) to train the model, it occurs this situation. I initially think there is something wrong with datasets (incomplete datasets), so I re-download the datasets and train again, the same evaluation results caused again. It seems the car only learn from datasets about how to turn a round. The only difference of my training configuration from the default configuration is that ‘--parallel_training’ we set ‘0’.

**Comment by Kait0** [comment]
Hm I doubt that the dataset is the issue as long as there are no corrupt file, and the folder root is placed correctly.
How many samples does the dataset contain once it is loaded?
We trained a lot of different models on this dataset and never encountered this problem.

In principle you don't need to change anything about the code to train on a single GPU, you just need to set --parallel_training to 0 and everything else should work fine.

How do your loss curves in tensorboard look like. Is the training stable?

Why did you change the schedule to
"schedule_reduce_epoch_01": 10,
"schedule_reduce_epoch_02": 20,
"schedule_reduce_epoch_03": 30,
?
reducing the learning rate once by a factor of 10 at epoch 30 worked well for us (and then usually we evaluated epoch 31 for ablations which worked well).
If you already reduce the learning rate by epoch 10 that might be too early.

You train with a different batch size (32). We usually used something like batch size 96 split across 8 GPUs.
This means your optimization does a lot more steps with noisier gradients. 
It's hard to tell what impact that has, maybe you can see it in the loss plots.


When you train with all the losses, you can use the [visualization code](https://github.com/autonomousvision/transfuser/blob/bef1fcf5622c3ab400d4c1f27059786e7cfecb11/team_code_transfuser/model.py#L935) to look at what the auxiliary losses are doing. You can use it by setting the save path and debug flag [in the forward pass](https://github.com/autonomousvision/transfuser/blob/bef1fcf5622c3ab400d4c1f27059786e7cfecb11/team_code_transfuser/model.py#L719).
Is semantic segmentation / depth producing meaningful predictions or did the entire model not converge?

**Comment by HelloWorld416** [comment]
Thank you for your timely reply!
1.After we load the dataset, there are 226115 samples in ‘train_set’ and 4850 samples in ‘val_set’.
2.There are two figures for two training loss: we set ‘all’ at the first time ans set ‘02_05withheld’ at the second time. When I use GTX 3090 GPU, 24G video memory allows me to set the maximum batch_size to 32.
Then, I will follow your advice to train again with visualization code.
<img width="590" alt="first_train-with_set_'all'" src="https://user-images.githubusercontent.com/90030184/193971017-2a48af6c-44dc-4549-8ddd-acb7ad5591a6.png">
<img width="591" alt="second_train-with_set_'02_05 '" src="https://user-images.githubusercontent.com/90030184/193971024-61cedf7d-1093-4b5e-a614-aca3c0a5b77d.png">

**Comment by Kait0** [comment]
So the first training clearly diverged after epoch 5.
The loss curve should never go up during training. 
Also the total loss is much higher than in the second plot.
Was the first training with auxiliary losses? ('only_wp'=0)
In general we observed that 'only_wp'=0 gives better results as you can see in the ablation section of the paper so I recommend that.

The second training loss looks more reasonable, did it inhibit the same problems?
Since you trained with 02_05withheld in the second run, you should have validation losses so you can look at how they are behaving as well.

Regarding the settings, in the paper we only use the ‘all’ setting.
The ‘02_05withheld’ setting is a feature I added for the release in case people want to evaluate a model on the LAV routes where you are not allowed to train on town 02 and 05.
The all setting overall gives the better models because it trains on more data.


One common trick to deal with different batch sizes used in the community is to change the learning rate proportionally if you change the batch size.
Since you have 3x smaller batch size the optimization will do 3x more steps (as you have the same amount of epochs).
So you can reduce the initial learning rate by 3x to get a similar results (and hopefully a smaller initial learning rate will help with your divergence problem).
So I would recommend trying to train with initial "lr": 0.000033, 'only_wp'=0 and setting 'all'

**Comment by HelloWorld416** [comment]
Thanks for your timely reply!
1.I used auxiliary losses (‘only_wp=0’) at first training, but set ‘--setting’ = ‘02_05..’ not ‘all’. I think this the reason why the training diverged after epoch 5.
2.The second training used no auxiliary losses (‘only_wp=1’) and set ‘--setting’ = ‘all’. It did not inhibit the same problem.
Then, I would try to train with initial "lr": 0.000033, 'only_wp'=0 and setting 'all'. By the way, have you tried to train the model with one GPU? Also I want to make sure whether the samples of dataset I used to train are correct and the dataset is the same with you? Anyway, thank you very much for your great help. I will keep going on finding out the problem.

**Comment by Kait0** [comment]
We haven't trained the model with a single GPU as like you said it takes a long time.
We usually train with 8 GPUs which converges < 1 day.
We used the exact same dataset as we released for all our experiments in the paper.
226115 samples is correct.

**Comment by HelloWorld416** [comment]
Sorry for late reply. I refer your suggestion and reset the training configuration. After 41 epochs training, we evaluate the model and find the same problem as previous we meet. I am very confused about it. I check the losses and find it is nomal and decreasing. I doubt that there was something wrong with coordination transformation between training and testing. 
<img width="578" alt="2022-10-17 10-27-33 的屏幕截图" src="https://user-images.githubusercontent.com/90030184/196077582-adc8ad04-1df7-4197-82a6-78af4ee9f162.png">
<img width="574" alt="2022-10-17 10-27-34 的屏幕截图" src="https://user-images.githubusercontent.com/90030184/196077597-14af1a3a-bafb-4fee-993d-9d22e49de74d.png">
<img width="577" alt="2022-10-17 10-28-47 的屏幕截图" src="https://user-images.githubusercontent.com/90030184/196077610-8516f867-48ae-41ab-ac1f-291348d2cc13.png">
<img width="581" alt="2022-10-17 10-29-02 的屏幕截图" src="https://user-images.githubusercontent.com/90030184/196077619-55eedf65-58ec-4af5-9a96-cc1460a0f322.png">
<img width="590" alt="2022-10-17 10-25-35 的屏幕截图" src="https://user-images.githubusercontent.com/90030184/196077627-c3f7ad95-60a1-4eb6-ae49-a7af0cf8fa7b.png">
Can I send the program and trained model to you to help me check the root cause of the problem？

**Comment by Kait0** [comment]
You can send it to me. When I have time i'll have a look.

**Comment by HelloWorld416** [comment]
That is very kind of you, thanks a lot. Can I have your email address?  or you can sent an email to me to lyh_0416@163.com, so that I can get contact with you.

**Comment by Kait0** [comment]
You can find it on my [website](https://kait0.github.io/).

**Comment by Kait0** [comment]
I had a look at your model.
The problem is that the model weights were not loaded correctly during inference.
When training a model with distributed data parallel the model weights are named slightly differently than when training with a single GPU.
For that reason we remove the added name when loading the model.
But because your model weights are trained with a single GPU they don't have that additional name portion and the code butchers the model names.
The result is that the model you evaluated was untrained.

To load your model weights correctly simply comment line 95 in [submission_agent.py](https://github.com/autonomousvision/transfuser/blob/bef1fcf5622c3ab400d4c1f27059786e7cfecb11/team_code_transfuser/submission_agent.py#L95)
If you want to make sure that everything went right you can also set loading strict=True in the following line.
After doing that change your trained model seemed to run fine at the short sequence I looked at.

**Comment by HelloWorld416** [comment]
Thanks for your sincere help! The problem has been solved. My own trained model has been worked.
